boards/DCM/Src/App/states/App_DriveState.c

```diff
@@ -1,20 +1,38 @@
+#include <stdlib.h>
+#include <math.h>
+#include "App_SharedMacros.h"
 #include "states/App_AllStates.h"
 #include "states/App_InitState.h"
 #include "App_SetPeriodicCanSignals.h"
+#include "configs/App_EfficiencyLUT.h"
 
-// TODO: Have separate maximum torque requests for motoring and generating
 #define MAX_TORQUE_REQUEST_NM (90.0f)
+#define RPM_TO_RADS(rpm) ((rpm) * (float)M_PI / 30.0f)
 
 void App_SetPeriodicCanSignals_TorqueRequests(struct DcmCanTxInterface *can_tx, struct DcmCanRxInterface *can_rx)
 {
-    float torque_request = 0.0f;
-
-    //              Accelerator Pedal Percentage
-    //  Torque =  -------------------------------  * MAX_TORQUE_REQUEST_NM
-    //                        100%
-    //
+    uint8_t lut_idx;
+    float   efficiency_estimate;
+    float   max_torque_request, torque_request;
+    float   bms_available_power = App_CanRx_BMS_AVAILABLE_POWER_GetSignal_AVAILABLE_POWER(can_rx);
+    float right_motor_speed_rpm = (float)abs(App_CanRx_INVR_MOTOR_POSITION_INFO_GetSignal_D2_MOTOR_SPEED_INVR(can_rx));
+    float left_motor_speed_rpm  = (float)abs(App_CanRx_INVL_MOTOR_POSITION_INFO_GetSignal_D2_MOTOR_SPEED_INVL(can_rx));
+
+    // Get the estimated efficiency for the current motor speed
+    lut_idx =
+        (uint8_t)min((uint8_t)right_motor_speed_rpm / EFFICIENCY_LUT_RPM_INTERVAL, EFFICIENCY_LUT_NUM_ENTRIES - 1);
+    efficiency_estimate = EfficiencyLUT[lut_idx].efficiency_estimate;
+
+    // Estimate the maximum torque request to draw the maximum power available from the BMS
+    // Note that the motors can not exceed a torque of MAX_TORQUE_REQUEST_NM
+    max_torque_request =
+        min(bms_available_power * efficiency_estimate /
+                (RPM_TO_RADS(right_motor_speed_rpm) + RPM_TO_RADS(left_motor_speed_rpm)),
```

Comment:
Kinda nit but can we create a new float variable to hold this value and pass it later into the min macro? Will be a bit nicer to read

Reply:
resolved